### DISCUSSION TITLE
RFC - API Tokens

### DISCUSSION THREAD

**Discussion by yuval-hazaz** [amplication/amplication#1228] [General] [answered]
# RFC - API Tokens

## Summary
Amplication server provides two authentication methods - OAuth with GitHub and a local account with a username (email) and password.
The hosted service is using OAuth with GitHub, while a self-hosted installation can use either method.
Also, additional authentication providers may be added in the future.

In order to use Amplication CLI (and maybe other tools in the future), the user needs to access Amplication Server API from a command-line interface. While it is possible to implement OAuth flow in a CLI, using an API Token is much simpler and common.

Also, using API token will provide a unified experience and a single code base that will support connecting to Amplication server with any of the current or future authentication providers.

This RFC provides a proposal and details on a method to implement support for API tokens. 

## Proposal
- Users will be able to generate API Tokens using Amplication UI.
- The API token will provide the same permissions and privileges as the user who created and own the token.
- A user can have multiple tokens.
- Each token will have an identifiable name.
- Tokens are valid for 30 days from creation or last use so that the 30-day expiration automatically refreshes with each API call.
- Tokens that aren't used for 30 days expire.
- A user will be able to deactivate or delete a token.

### Create Token
Below is a flow to create a new token:
- Signs in to the Amplicaiton server and navigate to an  "API tokens" page.
- See a list of API tokens that were previously created. 
- Click on a "Create API Token" button to create a new token.
- Provide a name for the new token.
- Click on "Create"
- See the full token. This is the only opportunity to see it and record it.

### Delete Token
Below is a flow to delete a token:
- Signs in to the Amplicaiton server and navigate to an  "API tokens" page.
- See a list of API tokens that were previously created. 
- For each token, show: name, last access date and time, and the last X characters of the token.
- Click on Delete
- The token will be deleted, and any client using this token will no longer be able to access the server.

### DB design
```
model ApiToken {
  id			String   @id @default(cuid())
  createdAt		DateTime @default(now())
  updatedAt		DateTime @updatedAt
  user			User     @relation(fields: [userId], references: [id])
  userId		String
  token			String		#Hashed String
  previewChars		String		#plain text string with last X characters of the token
  lastAccessAt		DateTime 
}
```

### Create Token API
Creates a token for the current user
The full token is not saved to DB.
Returns the full token to the client
```
createToken() : string  
```

### Get Tokens API
Returns the list of tokens for the current user
```
getTokens() : ApiToken[]
```

## Token Format
To keep using the existing validation mechanism in the server, JWT access token will be used to create the API tokens.
The JWT tokens will be created using the existing services that currently create the user JWT on the sign in, with an additional property to indicate that the token was created as an API key.
TBC: Can we use plain unique keys instead? 

## Token Client Usage 
The client must send the API token in the Authorization header when making requests to the server:
Authorization: Bearer <token>

## Token Validation
The server will validate the API key in the exact same way it already validates JWT, with an additional step to validate the token expiry and update the LastAccessAt field for the token.

**Comment by yuval-hazaz** [comment, marked as answer]
completed in v0.5.0

**Comment by cloudcompute** [comment]
Hello

I have few questions, could you pl. answer:
a. What is the difference between API tokens and JWT tokens?
b. JWT tokens should be stateless. If we are going to store them in a db, it hampers scalability. Why have you chosen to store them in the db?
c. These tokens are usually generated in pairs, it is probably because this pair a. Refresh tokens b. Access tokens .. makes statelessness possible.
Kindly have a look in the following starter kit 
https://github.com/fivethree-team/nestjs-prisma-starter/blob/4547666b8f95cfbd9b9600f7b093ec5cc61652c9/src/resolvers/auth/auth.resolver.ts#L29

Looks like Amplification is not generating this pair, am I right?

**Reply by cloudcompute** [reply]
One more point.. This pair, refresh and access token also probably makes sure to prevent CSRF and XSS attacks.

**Reply by yuval-hazaz** [reply]
Hi @ramandhingra , thanks for asking.

1. "API token" is just the name we choose for this type of token which can be used with the API without knowing the username and password, or more specifically without the link to GitHub login which we use in the hosted version. We use JWT as the format of the token.
2. JWT is stateless, or more specifically it does not require a state on the server to be used. But, also since it is stateless usually there is no way to invalidate a token once it was created and given to the user unless you validate the token with some kind of list (DB in this case) on the server-side. 
Since the API token usually are used to integrate between machine to machine (for example, in our case from a script against the CLI) they are commonly saved somewhere, and there is a need for central management and monitoring of the tokens, and also the option to invalidate a token once it is no longer needed. 
3. We do not use refresh tokens. Refresh tokens are part of the OAuth but are not a must for every use case with JWT. In our case, the API tokens are not used in an OAuth flow.
4. I am not sure why or how refresh tokens are used to prevent CSRF and XSS.

**Comment by cloudcompute** [comment]
Hi @yuval-hazaz 

I am not sure whether refresh tokens really help mitigate csrf and xss... I read this somewhere in a blog. I do not have complete knowledge how it actually works. But this is for sure, developers do make use of the pair (access and refresh tokens) for authentication.

I'd like to state one point.. If we store the JWTs at the server side, then there is no point in using the JWTs. Instead, we can use the battle-tested solution of storing the plain session ids at the server side.

http://cryto.net/~joepie91/blog/2016/06/13/stop-using-jwt-for-sessions/
